safetensors/src/slice.rs

```diff
@@ -0,0 +1,416 @@
+use crate::TensorView;
+use std::ops::{
+    Bound, Range, RangeBounds, RangeFrom, RangeFull, RangeInclusive, RangeTo, RangeToInclusive,
+};
+
+/// Error representing invalid slicing attempt
+#[derive(Debug)]
+pub enum InvalidSlice {
+    TooManySlices,
+}
+
+#[derive(Debug)]
+pub enum TensorIndexer {
```

Comment:
This is taken from https://github.com/LaurentMazare/tch-rs/blob/f691db1708c12db02f0836348f6ba341d3a0c40d/src/tensor/index.rs

I removed the unnecessary parts (for now) as it simplifies the code for now.

I'm leaving parts in comments because they might become important again (`[:, 2]` is `Select` for instance and could be useful as it would avoid a reshape, still unsure if I want to support here or in python only.
`IndexSelect` is ` [:, torch.Tensor(2,3 , 5, 6)]`  which I'm not sure we need/want support (first of all this exact form would require tensor support, as done in `tch-rs`. So heavy adaptation is probably needed.
In most cases in distributed settings I would say `tensor[:, ::3]` would be the most complex query and it would be written `(0..n).step_by(3)` in rust and is actually a `StepBy` object.

Anyway, all this complexity is left for later.